### PR TITLE
this will add completed task 3

### PR DESCRIPTION
**Pull request by WolfieKnee**
Completed task 3 for /api/endpoints.
Also addressing feedback on ticket 2.

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -15,28 +16,40 @@ beforeEach(() => seed(data));
...
+			.then((fileContents) => {
+				return JSON.parse(fileContents);
+			})
+			.then((expected) => {
+				return request(app)
+					.get("/api")
+					.expect(200)
+					.then((response) => {
+						const { endpoints } = response.body;
+						expect(endpoints).toMatchObject(expected);
```
Comment (alice-y-codes) [inline __tests__/app.test.js:30]: would like you to hardcode what the endpoint will look like instead of using fs.readfile because this test will auto pass for any subsequent updates

File: __tests__/app.test.js [hunk 2]
```diff
@@ -15,28 +16,40 @@ beforeEach(() => seed(data));
...
+	test("GET: 200 should respond with status code 200", () => {
+		return request(app).get("/api/topics").expect(200);
+	});
+	test("GET: should respond with an array of topic objects, each of which should have the following properties: slug, description", () => {
+		return request(app)
+			.get("/api/topics")
+			.expect(200)
+			.then((response) => {
+				const { topics } = response.body;
+				expect(Array.isArray(topics)).toBe(true);
```
Comment (alice-y-codes) [inline __tests__/app.test.js:45]: Please check length of the array back alongside a forEach as using only a .forEach will automatically pass the test

File: __tests__/app.test.js [hunk 3]
```diff
@@ -15,28 +16,40 @@ beforeEach(() => seed(data));
...
+				topics.forEach((topic) => {
+					expect(topic).toHaveProperty("slug", expect.any(String));
+					expect(topic).toHaveProperty(
+						"description",
+						expect.any(String)
+					);
 				});
-		});
+			});
 	});
```
Comment (alice-y-codes) [inline __tests__/app.test.js:54]: Please add a test for a potentially that may occur when you got to a path that does not exist

File: app.js [hunk 4]
```diff
@@ -1,13 +1,27 @@
...
 
 // TODO: when receiving a json body from the client, e.g. POST express.json()
 
 app.get("/api/topics", getTopics);
 
+app.get("/api", getEndpoints);
+
+// ERROR HANDLING
+app.use((err, req, res, next) => {
+	// handle error if endpoint.json file is missing. Cannot test for this as the filename is not given as part of endpoint, but manually confirmed in development with malformed path.
```
Comment (alice-y-codes) [inline app.js:15]: remove all comments and console.logs before pushing from now on

**Review by alice-y-codes — COMMENTED** [review]
see comments

**Comment by WolfieKnee** [comment]
Thanks Alice - a couple of those are consistent with feedback from Poonam (she didn't let them slip past!), but I lost the refactor by doing a PR wrong. I'll make sure to address these in a future PR.
Dave

File: __tests__/app.test.js (hunk 1 above)
Comment (WolfieKnee) [reply __tests__/app.test.js:30]: Since the content of endpoint.json is dynamic, and the only consistent key on each endpoint is "description" I have added a `for in` to test for this in each endpoint object. 
I have also included a test to verify that the object I am iterating over in non-zero.

File: __tests__/app.test.js (hunk 2 above)
Comment (WolfieKnee) [reply __tests__/app.test.js:45]: added a check for length

File: __tests__/app.test.js (hunk 3 above)
Comment (WolfieKnee) [reply __tests__/app.test.js:54]: added `app.all("/api/*"...` on app.js

File: app.js (hunk 4 above)
Comment (WolfieKnee) [reply app.js:15]: removed comments and consoles.
